ambry-router/src/main/java/com.github.ambry.router/GetBlobOperation.java

```diff
@@ -0,0 +1,807 @@
+  private final FirstGetChunk firstChunk;
+  // Associated with all data chunks in the case of composite blobs. Only a fixed number of these are initialized.
+  // Each of these is initialized with the information required to fetch a data chunk and is responsible for
+  // retrieving and adding it to the list of chunk buffers. Once complete, they are reused to fetch subsequent data
+  // chunks.
+  private GetChunk[] dataChunks;
+  // the factory to use to deserialize keys in a metadata chunk.
+  private final BlobIdFactory blobIdFactory;
+  // the total number of data chunks associated with this blob.
+  private int numChunksTotal;
+  // the total number of data chunks retrieved so far (and may or may not have been written out yet).
+  private int numChunksRetrieved;
+  // a list iterator to the chunk ids that need to be fetched for this operation, if this is a composite blob.
+  private ListIterator<StoreKey> chunkIdIterator;
+  // chunk index to retrieved chunk buffer mapping.
+  private Map<Integer, ByteBuffer> chunkIndexToBuffer;
+  // To find the GetChunk to hand over the response quickly.
+  private final Map<Integer, GetChunk> correlationIdToGetChunk = new HashMap<>();
+  // The result of this operation. This is instantiated if/when the first bytes of the result arrive and just before
+  // the operation callback is invoked.
+  private GetBlobResult getBlobResult;
+
+  private static final Logger logger = LoggerFactory.getLogger(GetBlobOperation.class);
+
+  /**
+   * Construct a GetBlobOperation
+   * @param routerConfig the {@link RouterConfig} containing the configs for get operations.
+   * @param routerMetrics The {@link NonBlockingRouterMetrics} to be used for reporting metrics.
+   * @param clusterMap the {@link ClusterMap} of the cluster
+   * @param responseHandler the {@link ResponseHandler} responsible for failure detection.
+   * @param blobIdStr the blob id associated with the operation in string form.
+   * @param futureResult the future that will contain the result of the operation.
+   * @param callback the callback that is to be called when the operation completes.
+   * @param operationCompleteCallback the {@link OperationCompleteCallback} to use to complete operations.
+   * @param time the Time instance to use.
+   * @throws RouterException if there is an error with any of the parameters, such as an invalid blob id.
+   */
+  GetBlobOperation(RouterConfig routerConfig, NonBlockingRouterMetrics routerMetrics, ClusterMap clusterMap,
+      ResponseHandler responseHandler, String blobIdStr, FutureResult<ReadableStreamChannel> futureResult,
+      Callback<ReadableStreamChannel> callback, OperationCompleteCallback operationCompleteCallback, Time time)
+      throws RouterException {
+    super(routerConfig, routerMetrics, clusterMap, responseHandler, blobIdStr, futureResult, callback, time);
+    this.operationCompleteCallback = operationCompleteCallback;
+    this.blobIdFactory = new BlobIdFactory(clusterMap);
```

Comment:
minor: curious if this can be passed from the `GetManager` (or whoever is creating this object) instead of creating a new object for every `GetBlobOperation`?